### PR TITLE
cmake: Replace Enable_PTHREAD with ENABLE_THREADS

### PR DESCRIPTION
**Pull request by e10harvey**
The merge of https://github.com/kokkos/kokkos/pull/4831 caused the KokkosKernels Pthread CI checks to start failing in https://github.com/kokkos/kokkos-kernels/pull/1355.

This MR adds the `Kokkos_ENABLE_THREADS` cmake option for users; I will update the kokkos-kernels CI checks to use `Threads` instead of `Pthread` in https://github.com/kokkos/kokkos-kernels/pull/1355

### CODE REVIEW

File: cmake/kokkos_tpls.cmake
Comment (masterleinad) [inline cmake/kokkos_tpls.cmake:63]: I thought @ndellingwood was saying that the `Thread` backed should bot be enabled/disabled based on `Trililinos`' `Pthread` TPL.

**Review by ndellingwood — APPROVED** [review]
Changes seem fine to me, thanks @e10harvey. I can test with Trilinos though this looks good, after approval+merge this will need a cherry-pick for the 3.6 release

File: cmake/kokkos_tpls.cmake
Comment (e10harvey) [reply cmake/kokkos_tpls.cmake]: @masterleinad : Ready to merge?

Comment (masterleinad) [reply cmake/kokkos_tpls.cmake]: I don't have the power to do that but I'm happy to approve (assuming @ndellingwood verified that this works correctly with `Trilinos`).

**Review by masterleinad — APPROVED** [review]

**Comment by ndellingwood** [comment]
@e10harvey @masterleinad There are a couple problems configuring Trilinos with these changes, and a separate pre-existing issue I didn't catch beforehand. Here's a summary:

* In a configuration with Serial only things look good - no regressions where a Serial build triggers a Threads build partially addressed by #4830 

* In a configure with `-DKokkos_ENABLE_THREADS=ON` there are two issues: 
  - `KOKKOS_ENABLE_THREADS` is not defined in `KokkosCore_config.h`
  - Configuration error
```
CMake Error at kokkos/cmake/kokkos_tpls.cmake:29 (MESSAGE):
    Enabled TPL THREADS inside TriBITS build, but this can only be enabled in a
      standalone build
```

* In a configure with deprecated `-DKokkos_ENABLE_PTHREAD=ON` there are two issues:
  - `KOKKOS_ENABLE_THREADS` is not defined in `KokkosCore_config.h`
  - The CMake cache variable `Kokkos_ENABLE_THREADS` is set OFF, but should be ON in this case - this was preexisting to these changes

**Comment by ndellingwood** [comment]
Regarding the final issue, does the FORCE option need to be added here https://github.com/kokkos/kokkos/blob/90fdc39dd357eb59efafc7b175e424e4d775c112/cmake/kokkos_enable_devices.cmake#L24 to ensure `Kokkos_ENABLE_THREADS=ON` when `Kokkos_ENABLE_PTHREAD=ON`?

**Comment by masterleinad** [comment]
Yes, I think so.

**Comment by masterleinad** [comment]
@ndellingwood Can you clarify if enabling according to `CMake` option in `Trilinos` is desired or not? Was that only an issue for the `Pthread`/`Thread` backend or would you anticipate that this is not desirable for other TPLs, too?
The `CMake` function currently enforces that only TPLs can be used that have an equivalent in `Trilinos`/`TriBITS` and that their value (in `Kokkos`) is set according to the `Trilinos` option.

**Comment by ndellingwood** [comment]
Making these change allowed the configurations of Trilinos to succeed as expected:

```diff
[ndellin@kokkos-dev cmake]$ git diff
diff --git a/cmake/kokkos_enable_devices.cmake b/cmake/kokkos_enable_devices.cmake
index 9bf8c98..dc3ee8c 100644
--- a/cmake/kokkos_enable_devices.cmake
+++ b/cmake/kokkos_enable_devices.cmake
@@ -21,7 +21,7 @@ KOKKOS_DEPRECATED_LIST(DEVICES ENABLE)
 
 KOKKOS_DEVICE_OPTION(THREADS OFF HOST "Whether to build C++ threads backend")
 IF(Kokkos_ENABLE_PTHREAD)  # for backward compatibility
-  SET(Kokkos_ENABLE_THREADS ON CACHE BOOL "Whether to build C++ threads backend")
+  SET(Kokkos_ENABLE_THREADS ON CACHE BOOL "Whether to build C++ threads backend" FORCE)
   SET(KOKKOS_ENABLE_THREADS ON)
   LIST(APPEND KOKKOS_ENABLED_DEVICES THREADS)
   SET(KOKKOS_HAS_HOST ON)
diff --git a/cmake/kokkos_tpls.cmake b/cmake/kokkos_tpls.cmake
index ddf7167..54c6b52 100644
--- a/cmake/kokkos_tpls.cmake
+++ b/cmake/kokkos_tpls.cmake
@@ -60,7 +60,7 @@ SET(HPX_DEFAULT OFF)
 ENDIF()
 KOKKOS_TPL_OPTION(HPX ${HPX_DEFAULT})
 
-KOKKOS_TPL_OPTION(THREADS ${Kokkos_ENABLE_THREADS})
+KOKKOS_TPL_OPTION(THREADS ${Kokkos_ENABLE_THREADS} TRIBITS Pthread)
 
 IF(Trilinos_ENABLE_Kokkos AND TPL_ENABLE_quadmath)
   SET(LIBQUADMATH_DEFAULT ON)
```

After the issues calling `KOKKOS_TPL_OPTION(THREADS ${Kokkos_ENABLE_THREADS} TRIBITS Pthread)` prior, I'm not sure what type of guards should be added...

**Comment by ndellingwood** [comment]
To try and be a bit more clear the issues I am referring to aren't specifically those of `KOKKOS_TPL_OPTION(THREADS ${Kokkos_ENABLE_THREADS} TRIBITS Pthread)` but the prior cmake function call for Pthreads re #4830 #4831

**Comment by ndellingwood** [comment]
@masterleinad I think the issue only came up with the Pthread/Threads backend because of the deprecation with renaming and something getting missed. More generally, I'm not sure for example if there are users (and how frequent) that set `TPL_ENABLE_CUDA=ON` and expect to get `Kokkos_ENABLE_CUDA=ON`

**Comment by masterleinad** [comment]
> @masterleinad I think the issue only came up with the Pthread/Threads backend because of the deprecation with renaming and something getting missed. More generally, I'm not sure for example if there are users (and how frequent) that set TPL_ENABLE_CUDA=ON and expect to get Kokkos_ENABLE_CUDA=ON

If we can, i.e., if that is what `Trilinos` wants, it's much easier to just delete https://github.com/kokkos/kokkos/blob/90fdc39dd357eb59efafc7b175e424e4d775c112/cmake/kokkos_tpls.cmake#L12-L18 and possibly https://github.com/kokkos/kokkos/blob/90fdc39dd357eb59efafc7b175e424e4d775c112/cmake/kokkos_tpls.cmake#L25-L31.

**Comment by ndellingwood** [comment]
> If we can, i.e., if that is what `Trilinos` wants, it's much easier to just delete

I'm not confident enough to give a thumbs up/down. For the example CUDA case I mentioned I didn't cross-check what was expected by TriBITS in terms of options (e.g. TPL_ENABLE_CUDA vs TPL_ENABLE_Cuda) and I thought I should mention that in case my scenario was misleading regarding importance/unimportance of case

**Comment by masterleinad** [comment]
We should just be aware that the last changes you posted imply setting `Kokkos_ENABLE_THREADS` according to `TPL_ENABLE_Pthread`. If that works fine, I'm happy with that.

**Comment by ndellingwood** [comment]
There's another Pthread/Threads linking problem I just filed an issue #4848, unrelated to this issue (introduced with #4831) but unresolved by the fixes in this PR

**Comment by ndellingwood** [comment]
Retested this PR and confirmed it resolved issues in #4848, I must not have cleaned the build before testing. So this PR will resolve that issue as well

**Review by ndellingwood — APPROVED** [review]

**Review by crtrott — APPROVED** [review]

File: cmake/Modules/FindTPLTHREADS.cmake
Comment (e10harvey) [reply cmake/Modules/FindTPLTHREADS.cmake:21]: Sure. FindTPLPTHREAD only searched for libpthread while FindTPLTHREAD searches for both libpthread and Threads and then links against either Threads or libpthreads depending on what was found to be available on the system.

Comment (masterleinad) [reply cmake/Modules/FindTPLTHREADS.cmake:21]: OK, I missed that `cmake/tpls/FindTPLPthread.cmake` already creates the TPL if `USE_THREADS` is false and `pthread.h` and `pthread` were found. We shouldn't need
```
 ELSE()
    KOKKOS_CREATE_IMPORTED_TPL(THREADS
            INTERFACE   #this is not a real library with a real location
            COMPILE_OPTIONS -pthread
            LINK_OPTIONS    -pthread)
```
here then, right?

Comment (e10harvey) [reply cmake/Modules/FindTPLTHREADS.cmake:21]: No, this is needed. It creates the THREADS cmake target for linking against libpthread.

**Pull request merged**
